### DISCUSSION TITLE
The new checkpoint listing in X/Y/Z Plot ...

### DISCUSSION THREAD

**Discussion by Woisek** [vladmandic/sdnext#618] [Ideas]
I have noticed just now, that when we do a x/y/z plot with checkpoints, the "copy button" now inserts "plates" (don't know how to call this) instead of a text list. With that, it's now possible to click and select specific checkpoints.
While this is certainly a very good idea for a few checkpoints or when being selective, it's a bit of a nightmare for collectors like me. It gets even more nightmarish when I want to do a checkpoint plot to test a prompt idea to see, which checkpoint is best suited for it.

What's the problem?
The problem is now ... the checkpoints are "plates" now and not a comma separated text list anymore.

Why is this a problem?
That's a problem because I have currently 176 checkpoints. From that 176 checkpoints there are 20 inpaint models, that I _don't_ want to include in the plot. So, I have to sort them out to have 156 potential checkpoints for a plot left.
Next problem is: I often have to decide if it should be a 1.5 or 2.1 model. I.e., when testing a LoRA made for 1.5, I can't include 2.1 models, because the generation would fail on the first 2.1 model it reaches. So I have to filter 2.1 out of the list.

For that, the filtering, I wrote myself a webpage that does this job for me. I can insert the whole string of checkpoints into a textfield, set filters and let it generate me a list.
BUT: I made it so, that I get a ... guess ... comma separated text list, that I could copy and paste into the x value of the x/y/z plot. But this is no more possible because ... guess again ... the checkpoints are now "plates" ... 😊😂

So ... to get to the point, is it somehow possible to change/add and:
1. get again a comma separated text list of all checkpoints and
2. use such list in the value field again

I don't know how others handle such workflow using multiple checkpoints in the X/Y/Z Plot, but for me, this would be somewhat important to have (again). And regarding those "plates" (I still don't know how to call them), in the case of having all 156 checkpoints activated doesn't make the UI very clear and pretty. But that's just by the way. 🙂

**Comment by teflocarbon** [comment]
You mean the hashes at the end of the model name? [83a817b8cd] as an example.
You don't need them, you can just remove them and use the regular model name. If you want an easy to parse list of models, you can find it under System Info. Just scroll down, open Models and they'll all be there. They're still comma separated, I'm not sure why you are seeing different behaviour for that.

**Reply by vladmandic** [reply]
x/y/z grid has been updated to use multi-select instead of using simple textbox to type values, so yes, what you're saying is correct. ideally, we should enable both multi-select and text, but gradio is limited in that way. i'll check.

**Reply by teflocarbon** [reply]
Ah, I see now. I haven't checked that since I updated the repo, whoops.
Multi-select is a lot more intuitive than text but as you said, ideally it would be best to support both since it makes it easier to send to others and save. Wouldn't be surprised if Gradio limits it though as you mentioned.

Perhaps this person is using the binder icon which is adding everything which is forcing them to 'sort it' by removing the ones they don't want?

**Comment by Woisek** [comment]
Additionally this one here:
This multi select box is crap (it was ever so). After the list is gone, I have no way to select additional elements, because the list doesn't appear. And with a shit load of checkpoints selcted (like I have 😋) this is the result:

![Zwischenablagebild](https://user-images.githubusercontent.com/122026301/235301018-018717cf-96de-4dcf-9df2-a371ddbccaad.png)

**Reply by teflocarbon** [reply]
Ah, I see the issue now. It looks like the dropdown box is being hidden behind the list of options after around 10+ checkpoints have been added. That's making it inaccessible unless you remove the checkpoints. I'm seeing this behaviour on my side at the moment.

**Comment by vladmandic** [comment]
@pangbo13 any chance we could make the new x/y/z script dropdowns disabled if list is larger then specific number of params? its problematic if list is really large. basically fallback to previous textbox. i really don't want to disable dropdown completely as in general i love them, just have them auto-disable for long lists?

**Reply by pangbo13** [reply]
> any chance we could make the new x/y/z script dropdowns disabled if list is larger then specific number of params?

Oh yes, we can achieve that, I'll work on it. But I'm a bit busy these two days, maybe next week I'll try to implement it.

**Reply by Woisek** [reply]
May I hook in? Would it be possible to have two inputs under each other? One is multi select, the other plain text. Depending on which box is filled, is used. Autopopulation would also only be done on the field where the according icon is clicked with the safety, that the other field would be automatically emptied and so no doubles can occur.

**Reply by Nem-z** [reply]
If I may: I have a little suggestion if possible. In the case you revert back to text after a certain number of checkpoint in the list. Would it be possible to add a text auto-complete feature for easier use?
I prefer the drop-down menu myself, but in case you need to revert back I think it could be useful to have an auto complete feature.
Of course, I don't want to sound ungrateful and I don't know how hard it is to apply add such feature.So, thank you for your time and work anyway.

**Reply by pangbo13** [reply]
@Woisek 

Thank you so much for your help! Sounds like a great idea, the only problem is that I'm a bit concerned if it's going to mess up the UI. But at least it's better than an overly long dropdown :D

**Reply by vladmandic** [reply]
having multiple inputs for the same thing makes extremely busy ui even busier, that would be my least preferred option.

**Reply by pangbo13** [reply]
Maybe a checkbox that allows the user to choose whether to enable the dropdown would be enough to solve most problems?

**Reply by vladmandic** [reply]
sure, thats an easy fix, but the problem with checkbox is its all-or-nothing.
for example, i'd like to have multi-select for samplers (short list), but not for models (very long list). if there are no better options, i'd approve any prs that introduce that checkbox.

**Reply by Woisek** [reply]
@vladmandic I see your concern about this "all-or-nothing" option (nice Milli Vanilli vibe 😊).
But maybe it's easier to decide in regards on how likely it will be, that 2 or more options with "extra long pre-fill" possibilites will be used. "Samples" is manageable, "styles" could also be an issue with a long list. But other than that, there should be nothing that produces such a long list.
On the other hand, when I test TI embeddings, it's not uncommon that for a S/R field I enter something like:
embedding-100, embedding-200, embedding-300, embedding-400, embedding-500, embedding-600, embedding-700, embedding-800, embedding-900, embedding-1000, embedding-1100, embedding-1200, embedding-1300, embedding-1400, embedding-1500, embedding-1600
Often even longer. And this is also no problem. At least for me. 🙂
I think, a checkbox sounds to be a great middle way. Better than my suggestion and less confusing. 😋

**Comment by vladmandic** [comment]
i just did quick&dirty - if choices are more than dozen, it will auto-disable and fallback to textbox.
if someone wants to submit a PR to add an on/off checkbox, i'm ok with that as well.

**Reply by Woisek** [reply]
What about the possibility to get a complete list of all checkboxes like before? Or happens this automatically when populated with all checkpoints?

**Reply by vladmandic** [reply]
for now, its one or the other, cant have both :)
if you ask @pangbo3 nicely, maybe he'll add it in the future?

**Reply by Woisek** [reply]
> for now, its one or the other, cant have both :)

But when it changes after an amount of entries (maybe 6 would be a nice value), this should also happen when auto filled with more than x amount of entries ... ?!
And I hope that @pangbo13 is already convinced about this and is already striving for a solution, right? 😁

**Reply by vladmandic** [reply]
what i've added is if number of choices is more than 12, use textbox and do not attempt to use multiselect.

autofill fills multi-select component, it doesn't auto-fill textbox. so it would be more coding - like if choices exist and more than 12, then use textbox, but add a new fill that fills that instead of multiselect. but if there are no choices, then don't show auto-fill. etc... it gets messy fast. 

i've done what i can for short term, really need to move to other things - and if @pangbo13 (or anyone else) submits a PR with improvements, great.

**Reply by pangbo13** [reply]
These functions are all achievable, but different people have different needs, and it is difficult to find a method that satisfies everyone. As far as I am concerned, I am willing to sacrifice the simplicity of the UI so that I can get more comprehensive control, while some people want the operation to be as simple as possible.

By the way, I'm a little confused 

> list of all checkboxes like before

what does "checkboxes" mean here?

**Reply by vladmandic** [reply]
> These functions are all achievable, but different people have different needs, and it is difficult to find a method that satisfies everyone.

yup, totally agree.

> As far as I am concerned, I am willing to sacrifice the simplicity of the UI so that I can get more comprehensive control, while some people want the operation to be as simple as possible.

I'm primarily for simplicity unless it removes some critical functionality.

> what does "checkboxes" mean here?

I think he'd like to have auto-fill available for textbox as well as checkbox controls.

**Reply by Woisek** [reply]
Um ... no, the checkbox refers to @pangbo13 comment:

>Maybe a checkbox that allows the user to choose whether to enable the dropdown would be enough to solve most problems?

And I thought, @vladmandic will implement it like this 

> you have persuaded me, checkbox it will be

**Reply by vladmandic** [reply]
> you have persuaded me, checkbox it will be

i've removed that comment 1min after i wrote it, amazing how you found it. i took another route instead - autodisable if list is longer than 12 entries.

**Reply by Woisek** [reply]
But wouldn't the checkbox be much simpler and clearer to the User ... ? I really liked this kind of suggestion, because no matter what the user prefers, he can choose either this or that.
If one wants the text box, he has now have to enter 13 random entries before he get to the text field and can paste in what he originally wanted to enter ... ?! Hmm ...

**Comment by Woisek** [comment]
Additional: And the other side of Users? Those who like to clicki clicki, the click i.e. the checkpoints and suddenly, after the 12th it switches to a text field and suddenly they now have nothing more to click on? How do they add more checkpoint then? 🙂

**Reply by vladmandic** [reply]
yes, i agree, i said so - having both would be great - but for now i really really need to focus on some other things. 
and i'm open to prs for sure.

**Reply by Woisek** [reply]
OK, to finish it for now: What is the last status you did with it?

**Reply by vladmandic** [reply]
if choices are longer than 12 do not use multiselect and fallback to use textbox.

**Reply by Woisek** [reply]
Ok ... I made the update and now I have only a textbox. No drop down, no button to auto-populate ... 😶
16:16:46-134958 INFO     Version: 79c7e9d0 Sat Apr 29 10:58:06 2023 -0400

Also, in the img2img tab, the CFG Scale and Denoising strength sliders are suddenly at the very bottom ... after the scripts ... ?

![Zwischenablagebild](https://user-images.githubusercontent.com/122026301/235311350-94c75f59-87ab-47ac-a625-e44d0565283b.png)

**Reply by Nem-z** [reply]
No drop-down or auto complete on the textbox on my side either.
`Version: 69ee51f7 Sat Apr 29 14:16:39 2023 -0400` `Windows 11` `Firefox` (tested without any extensions enabled too).

Steps taken:
- Open WebUI.
- Go to `img2img` or `txt2img`.
- Activate `X / Y / Z plot` script.
- Select `Checkpoint Name` in the `Type` menu.
- There is a small loading before the textbox on the right appear but after that no drop-down or autocomplete and even typing the checkpoints name manually does nothing.
- Error message in console: 
```
X/Y/Z plot will create 0 images on 1 0x1 grid; 4 images per cell. (Total steps to process: 0)
Unexpected error: Processing could not begin, you may need to refresh the tab or restart the service.
```
____
I do not have the CFG and Denoising sliders issue though. I'm using the `nota-ai` theme

**Reply by vladmandic** [reply]
But that's the whole point?!

If you select something with short list of choices, like samplers, you get multiselect pull down.
If you select something with long like like 100s of checkpoints, there is no pull down or multiselect, it falls back to simple textbox.

That is what this discussion was created for!
And then it became "I want texbox but still with autocomplete" and many others. And I'd welcome PRs for that, but otherwise, there will be no work to improve.

If there are issues, please do report.

**Reply by Nem-z** [reply]
Sorry I misunderstood the process... I diminished the number of checkpoints in my folder for testing and it goes back to drop-down and then it works again. The problem persist with textbox though...

I tried:
`modelname`
`modelname.safetensors`
`modelname.safetensors [hash]`

What am I missing?

**Reply by uwings** [reply]
> No drop-down or auto complete on the textbox on my side either. `Version: 69ee51f7 Sat Apr 29 14:16:39 2023 -0400` `Windows 11` `Firefox` (tested without any extensions enabled too).
> 
> Steps taken:
> 
> * Open WebUI.
> * Go to `img2img` or `txt2img`.
> * Activate `X / Y / Z plot` script.
> * Select `Checkpoint Name` in the `Type` menu.
> * There is a small loading before the textbox on the right appear but after that no drop-down or autocomplete and even typing the checkpoints name manually does nothing.
> * Error message in console:
> 
> ```
> X/Y/Z plot will create 0 images on 1 0x1 grid; 4 images per cell. (Total steps to process: 0)
> Unexpected error: Processing could not begin, you may need to refresh the tab or restart the service.
> ```
> 
> I do not have the CFG and Denoising sliders issue though. I'm using the `nota-ai` theme

i have this issue too , i can't fix ... becuase there is no more log , but i have many ckpts, maybe about... 100?

**Reply by vladmandic** [reply]
i cannot reproduce and i have 200+ checkpoints - drop down menu works fine.
note its really hard to track any issues if you comment in the middle of a discussion that did not have activity in weeks.

**Comment by vladmandic** [comment]
> Also, in the img2img tab, the CFG Scale and Denoising strength sliders are suddenly at the very bottom ... after the scripts ... ?

i cannot reproduce that.

**Reply by Woisek** [reply]
Strange ... 😐

**Comment by Woisek** [comment]
@vladmandic 
i think, there is some clarification needed.

>If you select something with short list of choices, like samplers, you get multiselect pull down.

Even with the latest update: no, it's just a plain textbox.
21:09:05-840954 INFO     Version: 3c410561 Sat Apr 29 16:04:10 2023 -0400
![Zwischenablagebild](https://user-images.githubusercontent.com/122026301/235322808-643abd71-ec4c-4049-9fd8-7b68fc0b2010.png)

>If you select something with long like like 100s of checkpoints, there is no pull down or multiselect, it falls back to simple textbox.

That is currently the case, but there is not button to fill it with all the available checkpoints. How do we know, what to enter? That's why the pre-fill button is for.

> That is what this discussion was created for!

Yes, but it seems it somehow got a wrong turn. 🙂

>And then it became "I want texbox but still with autocomplete" and many others. And I'd welcome PRs for that, but otherwise, there will be no work to improve.

No, no "autocomplete". It was a question on how to combine the advantages of both textbox and drop down. I suggested to display both and @pangbo13 suggested a checkbox as a switch between both. Or better said, switch the functionality of it. I still like this idea and think it would be a good compromise to satisfy both kind of Users: the ones who like to click and the ones who like to write.
Checkbox on: drop down, checkbox off: text box. And on the right Still the icon for the prefill of values for this selection. 

>If there are issues, please do report.

Like I said, still there is no drop down and icons anymore, only textfields.
The problem with the moved CFG slider is gone, it's in place again.

**Comment by Seafan75** [comment]
Got the latest commit and restarted/reloaded the page.
The issue is that the format for checkpoints is unclear (or it is just not working).

![image](https://user-images.githubusercontent.com/132169108/235323908-b00de3cd-3567-4f44-a9a0-b45d6ebc3b91.png)
![image](https://user-images.githubusercontent.com/132169108/235323920-9b868fd3-9901-4863-8c4d-c53497f3f318.png)

These all gives:
![image](https://user-images.githubusercontent.com/132169108/235323935-37f8a118-b527-45cd-848a-c7e8315a723f.png)

On chrome.

**Comment by litrk** [comment]
it was a very cool feature, it's a pity that now it is gone, making the use of XYZ in this ui impossible for me. I hope you reconsider your decision or at least raise the threshold to 50 checkpoints/styles. Now I have 36 checkpoints and over 40 styles (artists recorded as styles) and on my normal sized monitor I haven't run into any problems. I would even like to have more spacing. I'm using the flamingo_dark theme, maybe it makes lists more compact and easier to use...

**Comment by teflocarbon** [comment]
I can't reproduce any of the issues here. The dropdown menu no longer overlaps and moves down as more are added. The dropdown works fine and text works fine as well.
I don't see a limit on checkpoints or styles? I have a ton of styles and they just become scrollable as expected.

**Reply by Woisek** [reply]
I have Version: 67c39202 Sun Apr 30 11:16:58 2023 +0900 and the issue is, that there are no dropdown menus ...

**Reply by teflocarbon** [reply]
My hash is identical and there is a dropdown menu.
updated: 2023-04-30
hash: 67c39202
url: https://github.com/vladmandic/automatic.git/tree/master

Have you tried turning off your custom theme or disabling extensions?

**Reply by Nem-z** [reply]
@teflocarbon 
Are you testing in the `X / Y / Z plot` script?
![image](https://user-images.githubusercontent.com/63961359/235361542-7a4f8e29-e794-44b7-8896-dd50e4816411.png)

Since I used less than 12 Samplers I have access to the drop-down menu:
![image](https://user-images.githubusercontent.com/63961359/235361595-ab85f382-d331-419f-a879-58040a2c923b.png)

But I have more than 12 checkpoints so it obviously switched back to a simple textbox as mentioned above. The problem is that even if I write down the checkpoint name the process doesn't go through but if I use the drop-down it works as intended.
Tested with:
`modelname`
`modelname.safetensors`
`modelname.safetensors [hash]`

There error message:
```
X/Y/Z plot will create 0 images on 1 0x1 grid. (Total steps to process: 0)
Unexpected error: Processing could not begin, you may need to refresh the tab or restart the service.
```

**Reply by teflocarbon** [reply]
Yes, I'm testing in the same script.
I don't have a dropdown for samplers, but I have one for checkpoints.

**Reply by Woisek** [reply]
I don't use a custom theme and what extension could interfere with this core element of Gradio? The model selection dropdown is working, the sampler dropdown is working, every other dropdown in the extensions is working, so ...

**Reply by mitien** [reply]
Just tested fresh installation: threshold is 12 checkpoints. 
less than 12 - drop down and autofill with chips, more - plain text field.

Can it be done like : 
checkpoints count < threshold  =  dropdown + autofill with chips
checkpoints count >= threshold = text field + text autofill with comma separator like was in original automatica1111?

**Reply by Woisek** [reply]
The actual status here is as follows:
Selection X/Y/Z Prompt, select X type as "Checkpoint name"

Result:
X value field (#script_txt2txt_xyz_plot_x_values) is a textbox, a dropdown (no ID given) is hidden

Issue 1: The icon for auto-populate the box with all available checkpoints is hidden. It should be visible, all the time.
Issue 2: Making this icon show again and clicking on it popluates the dropdown, but not the textbox

**Reply by Woisek** [reply]
> Can it be done like : checkpoints count < threshold = dropdown + autofill with chips checkpoints count >= threshold = text field + text autofill with comma separator like was in original automatica1111?
Or simpler:
pangbo13
Apr 29, 2023
Maybe a checkbox that allows the user to choose whether to enable the dropdown would be enough to solve most problems?
🙂

**Reply by mitien** [reply]
I agree. It might be simpler but there is not enough space in the UI for that checkbox directly in withing scripts/
![image](https://user-images.githubusercontent.com/672923/235363567-beed704f-a46e-40c2-87c6-950f195e0c07.png)
So only option is to add it to settings where it might get lost among hundreds of another options.

**Reply by Woisek** [reply]
> I agree. It might be simpler but there is not enough space in the UI for that checkbox directly in withing scripts/

Well, it would be if the UI wouldn't have been changed to be smaller but stayed at half/half just as before ...

> So only option is to add it to settings where it might get lost among hundreds of another options.

No, this is no option that you set and forget. This is probably used almost every time depending on need. It can't be in the settings.

**Comment by Metachs** [comment]
The issue with the Dropdown Option List not showing up is caused by the Gradio Dropdown assuming it would never be taller than the viewport, and setting garbage values when it gets too tall.

This is trivial to work around by limiting the height of the Dropdown to a sane value, which it probably should be anyhow. For example, the following CSS in my user.css fixes the problem by limiting the Dropdown multiselect height to 50% of the viewport, and configuring it to scroll. 

```
.gradio-dropdown.multiselect div.wrap-inner {	
    overflow-x: hidden;
    overflow-y: auto;
    max-height: 50vh;
    overflow-wrap: anywhere;
}
``` 

IMO something like this a better solution than the current item limit, a limit of 12 items is too small, it doesn't even let you select from all the available Samplers, which is one place I don't think anyone would object to having an option list.

Also, I think it would make much more sense if the XYZ values were on a separate line from the XYZ type, even for entering text, the current width is way too small.

![image](https://user-images.githubusercontent.com/9453185/235375688-720912ca-c766-4ae0-b9e1-d2475de29f6c.png)

![image](https://user-images.githubusercontent.com/9453185/235375629-7dffdae3-0eb7-4af5-b4b7-faa2bcf9dff0.png)

**Comment by vladmandic** [comment]
i've reverted the change as it caused side-effects and added gradio style as suggested.
anything better, i'm open to PRs.

**Comment by Woisek** [comment]
I have to pick this up again. 😞
First, regarding better option: I'm still convinced that a added checkbox, that toggles the behavior between textfield and dropdown, would be the most elegant solution, because it's User driven.

But why I'm here: I (again) did a checkpoint preview of a prompt and noticed, that the list I had selected is partially ignored. Ignored in that matter, that I got grids of checkpoints, I didn't selected. See this screenshot:

![Zwischenablagebild](https://user-images.githubusercontent.com/122026301/236150129-a5d569b6-1cf9-40b6-9742-f5503f9f47cd.jpg)

You see, that, although vectorart_beta_SD2.1 768 is not selected in the list, it still rendered a grid from it.
It's not visible in this screenshot, but this wasn't the only one. I noticed that only, because there where around 4 to 6 other checkpoints, I haven't checked for plotting.

Anyone other experienced this behavior?

**Comment by Woisek** [comment]
No one knows about this? @vladmandic

**Reply by vladmandic** [reply]
Doesn't ring a bell.
If it's reproducible, can you create an issue for it?

**Reply by Woisek** [reply]
I will make a controlled test and report then.

**Comment by Woisek** [comment]
OK, I did a test now and it turned out ... strange, again. Look at this:
![Zwischenablagebild](https://user-images.githubusercontent.com/122026301/236504614-0b30de41-61a6-4240-b400-cac5484600df.jpg)

What you see here is the result of what should be a 23 checkpoint selected test.
First issue: There are only 22 folders (instead of 23).
Second: On the left side of the ss, the image in the center (left of the only b&w image) is wrong. It should be Hard_Steel_Merge_v1.0 but it used robots_true assistant v21 (on the right side of the ss, the folder in the lower left).
Third: Not only is it a wrong checkpoint, it also rendered 2 images of it. 😶

This is an other behavior I experienced before, but none the less a bit scary ... 😐
I wonder if that has something to do with the drop down, sending wrong values ... 🤨

**Comment by vladmandic** [comment]
Please create issue with this information so it doesn't fall of my radar.

**Reply by Woisek** [reply]
Is it OK if I link this thread to the issue, to not write it all again? 😋

**Comment by Woisek** [comment]
So, after some time gone ... can we discuss this "issue" again?
I'm doing more prompt/checkpoint evaluating again lately and it's so incredible tedious (and time consuming if one has to do this multiple times a day) always to off-click 20 to 40 from now whopping 175 checkpoints ... 😔

So I want to again suggest three options:

1. A checkbox besides/underneath the auto-fill icon, that switches between a text field and the dropdown (since the text field is already/still there but hidden, this should be possible). I still think that this is the most convenient option to give the user a controlled possibility to switch for what he prefers the most.

2. An option to switch between text field and dropdown for the X/Y/Z plot in the settings. Basically the in 1 mentioned checkbox, but stored in the settings. Not so convenient as 1 but I'm assuming that the user have their preferences and will probably not switch that often so this could be justified.

3. Make one dropdown a text field, i.e. the Z values (X and Y are left as dropdown). The idea would be, that it is highly likely, that not all three values are used with pre-fill options, so we could have at least one entry to use as a text filed where we can edit the values and then swap it to the plot we want.

As an example: I want to do an prompt/checkpoint plot, so I leave X empty, Y is a given seed and in Z I copy and paste my selected checkpoints. Then I swap X/Z axis and I can do my plot. I know this is very unconventional, but at least it would be possible.

So, what do you think? 🙂

**Reply by vladmandic** [reply]
feature request #778 is open and i'd love contributions. i'll add notes there.

**Reply by Woisek** [reply]
Oh, great! Thanks for looking into it again. 👍👍
